File: option1.js
import React, { useState } from 'react';
import axios from 'axios';

function Backtester() {
  const [strategy, setStrategy] = useState('');
  const [portfolio, setPortfolio] = useState('');
  const [startDate, setStartDate] = useState('');
  const [endDate, setEndDate] = useState('');
  const [result, setResult] = useState('');

  const handleStrategyChange = (event) => {
    setStrategy(event.target.value);
  };

  const handlePortfolioChange = (event) => {
    setPortfolio(event.target.value);
  };

  const handleStartDateChange = (event) => {
    setStartDate(event.target.value);
  };

  const handleEndDateChange = (event) => {
    setEndDate(event.target.value);
  };

  const handleBacktest = () => {
    axios.post('http://localhost:5000/backtest', {
      strategy: strategy,
      portfolio: portfolio,
      start_date: startDate,
      end_date: endDate,
    })
    .then((response) => {
      setResult(response.data);
    })
    .catch((error) => {
      console.error(error);
    });
  };

  return (
    <div>
      <h1>Backtester</h1>
      <form>
        <label>Strategy:</label>
        <select value={strategy} onChange={handleStrategyChange}>
          <option value="">Select a strategy</option>
          <option value="mean_reversion">Mean Reversion</option>
          <option value="moving_average_crossover">Moving Average Crossover</option>
        </select>
        <br />
        <label>Portfolio:</label>
        <select value={portfolio} onChange={handlePortfolioChange}>
          <option value="">Select a portfolio</option>
          <option value="sp500">S&P 500</option>
          <option value="dow_jones">Dow Jones</option>
        </select>
        <br />
        <label>Start Date:</label>
        <input type="date" value={startDate} onChange={handleStartDateChange} />
        <br />
        <label>End Date:</label>
        <input type="date" value={endDate} onChange={handleEndDateChange} />
        <br />
        <button onClick={handleBacktest}>Backtest</button>
      </form>
      <h2>Result:</h2>
      <p>{result}</p>
    </div>
  );
}

export default Backtester;

File: option1.py
from flask import Flask, request, jsonify
import pandas as pd
import numpy as np

app = Flask(__name__)

# Load historical data
historical_data = pd.read_csv('historical_data.csv')

# Define strategies
strategies = {
    'mean_reversion': mean_reversion_strategy,
    'moving_average_crossover': moving_average_crossover_strategy,
}

# Define portfolios
portfolios = {
    'sp500': sp500_portfolio,
    'dow_jones': dow_jones_portfolio,
}

def mean_reversion_strategy(data, portfolio):
    # Implement mean reversion strategy
    pass

def moving_average_crossover_strategy(data, portfolio):
    # Implement moving average crossover strategy
    pass

def sp500_portfolio(data):
    # Implement S&P 500 portfolio
    pass

def dow_jones_portfolio(data):
    # Implement Dow Jones portfolio
    pass

@app.route('/backtest', methods=['POST'])
def backtest():
    strategy = request.json['strategy']
    portfolio = request.json['portfolio']
    start_date = request.json['start_date']
    end_date = request.json['end_date']

    # Filter historical data
    data = historical_data[(historical_data['date'] >= start_date) & (historical_data['date'] <= end_date)]

    # Run strategy
    result = strategies[strategy](data, portfolios[portfolio])

    return jsonify({'result': result})

if __name__ == '__main__':
    app.run(debug=True)
